### PR TITLE
#2022 spark-commons

### LINKED ISSUE #2022: Use spark-commons library

### PR DESCRIPTION
**Pull request by AdrianOlosutean**
Closes #2022

### CODE REVIEW

File: spark-jobs/src/main/scala/za/co/absa/enceladus/common/CommonJobExecution.scala [hunk 1]
```diff
@@ -326,11 +326,13 @@ trait CommonJobExecution extends ProjectMetadata {
...
   protected def addInfoColumns(intoDf: DataFrame, reportDate: String, reportVersion: Int): DataFrame = {
-    import za.co.absa.enceladus.utils.implicits.DataFrameImplicits.DataFrameEnhancements
+    import za.co.absa.spark.commons.implicits.DataFrameImplicits.DataFrameEnhancements
+    val function: (DataFrame, String) => DataFrame = (df: DataFrame, _) =>
+      df.withColumn("errCol", lit(Array.emptyIntArray))
     intoDf
-      .withColumnIfDoesNotExist(InfoDateColumn, to_date(lit(reportDate), ReportDateFormat))
-      .withColumnIfDoesNotExist(InfoDateColumnString, lit(reportDate))
-      .withColumnIfDoesNotExist(InfoVersionColumn, lit(reportVersion))
+      .withColumnIfDoesNotExist(function)(InfoDateColumn, to_date(lit(reportDate), ReportDateFormat))
```
Comment (dk1844) [inline spark-jobs/src/main/scala/za/co/absa/enceladus/common/CommonJobExecution.scala:333]: It is really late, but I think I take issue with `DataFrameImplicits.DataFrameEnhancements.withColumnIfDoesNotExist(ifExists: (DataFrame, String) => DataFrame)(colName: String, colExpr: Column)` that originated in https://github.com/AbsaOSS/spark-commons/pull/18, apparently.

IMHO the name of the method simply does not conform to what it does and what params it requires - the name suggests no-op column does not exist, but it allows an action in such a case, too. My vote is either to have two separate methods `withColumnIfDoesNotExist` and `transformIfColumnExists` or at least rename this method somehow to make sense.

Here, you can alleviate the problem only by renaming the `function` to something more descriptive like `ifExistsFn` or similar. (this occurs in multiple codebase files)

File: spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/interpreter/stages/TypeParser.scala [hunk 2]
```diff
@@ -194,7 +197,7 @@ object TypeParser {
       logger.info(s"Creating standardization plan for Array $inputFullPathName")
       val origArrayType = origType.asInstanceOf[ArrayType] // this should never throw an exception because of `checkSetupForFailure`
       val arrayField = StructField(fieldInputName, fieldType.elementType, fieldType.containsNull, field.structField.metadata)
-      val lambdaVariableName = s"${SchemaUtils.unpath(inputFullPathName)}_${Random.nextLong().abs}"
+      val lambdaVariableName = s"${za.co.absa.enceladus.utils.schema.SchemaUtils.unpath(inputFullPathName)}_${Random.nextLong().abs}"
```
Comment (dk1844) [inline spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/interpreter/stages/TypeParser.scala:200]:
```suggestion
      import za.co.absa.enceladus.utils.schema.{SchemaUtil => EnceladusSchemautils} // import rename
      val lambdaVariableName = s"${EnceladusSchemautils.unpath(inputFullPathName)}_${Random.nextLong().abs}"
```

**Review by dk1844 — DISMISSED** [review]
LGTM
- [x] code reviewed
- [x] pulled
- [ ] built
- [ ] run

File: spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/HyperConformance.scala [hunk 3]
```diff
@@ -78,11 +78,13 @@ class HyperConformance (menasBaseUrls: List[String],
     // using HDFS implementation until HyperConformance is S3-ready
     implicit val hdfs: FileSystem = FileSystem.get(sparkSession.sparkContext.hadoopConfiguration)
     implicit val hdfsUtils: HadoopFsUtils = HadoopFsUtils.getOrCreate(hdfs)
+    val function: (DataFrame, String) => DataFrame = (df: DataFrame, _) =>
```
Comment (dk1844) [inline spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/HyperConformance.scala:81]: that non-descriptive name of function-val you already fixed elsewhere -- here, you might have forgotten to change it, too.

**Review by dk1844 — APPROVED** [review]
LGTM

File: spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/HyperConformance.scala (hunk 3 above)
Comment (AdrianOlosutean) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/HyperConformance.scala:81]: I noticed a mistake. It's not only the naming that I did wrong here

**Review by dk1844 — DISMISSED** [review]

File: utils/src/main/scala/za/co/absa/enceladus/utils/schema/SparkUtils.scala [hunk 4]
```diff
@@ -104,4 +86,7 @@ object SparkUtils {
     dfWithAggregatedErrColumn.drop(tmpColumn)
   }
 
+  def ifExistsErrorFunction(colExpr: Column): (DataFrame, String) => DataFrame =
```
Comment (benedeki) [inline utils/src/main/scala/za/co/absa/enceladus/utils/schema/SparkUtils.scala:89]: I think the function name is more confusing than helping, took my a while to understand. 
I would prefer using the `overwriteWithErrorColumn(_, _, colExpr)` as partially defined.
And base on the observation of usage perhaps create a function replacing `withColumnIfDoesNotExist(overwriteWithErrorColumn)...` with something like `withColumnOverwrightIfExists(colName, colExpr)`.

File: spark-jobs/src/main/scala/za/co/absa/enceladus/common/CommonJobExecution.scala (hunk 1 above)
Comment (benedeki) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/common/CommonJobExecution.scala]: I tend to agree. In minimum, the parameters order seems wrong, and could have a default value in the `ifExists` parameter.

File: utils/src/main/scala/za/co/absa/enceladus/utils/implicits/StructFieldImplicits.scala [hunk 5]
```diff
@@ -1,47 +0,0 @@
...
- * See the License for the specific language governing permissions and
- * limitations under the License.
- */
-
-package za.co.absa.enceladus.utils.implicits
-
-import org.apache.spark.sql.types._
-import scala.util.Try
-
-object StructFieldImplicits {
```
[the remark this answers is not in the document]
Comment (AdrianOlosutean) [reply utils/src/main/scala/za/co/absa/enceladus/utils/implicits/StructFieldImplicits.scala:21]: They are already in spark-commons, just a bit renamed and on Metadata. I'm not sure if it makes sense to be duplicated in Enceladus and spark-commons

File: utils/src/main/scala/za/co/absa/enceladus/utils/schema/SparkUtils.scala (hunk 4 above)
Comment (AdrianOlosutean) [reply utils/src/main/scala/za/co/absa/enceladus/utils/schema/SparkUtils.scala]: Done

File: utils/src/main/scala/za/co/absa/enceladus/utils/implicits/StructFieldImplicits.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (benedeki) [reply utils/src/main/scala/za/co/absa/enceladus/utils/implicits/StructFieldImplicits.scala:21]: Hm, I see. They are useful, but I like them even better in the original form, being more concise. 🤷‍♂️

**Comment by benedeki** [comment]
Refactoring, doesn't really need/can be retested.

**Review by benedeki — APPROVED** [review]
Code looks fine, but my build *unit tests) fails, while `develop-ver-3.0` passes.

**Pull request merged**